### DISCUSSION TITLE
PropEr installation doesn't follow Erlang application structure

### DISCUSSION THREAD

**Discussion by mezzomondo** [Homebrew/discussions#2537] [Casks]
TL;DR: is it possible for `brew` to install `proper` in a folder named for example `/opt/homebrew/Cellar/proper-1.4` instead of `/opt/homebrew/Cellar/proper/1.4`?

The installation of proper (with `brew install proper`) creates the following directory structure:

```
/opt/homebrew/Cellar/proper/1.4
├── COPYING
├── INSTALL_RECEIPT.json
├── README.md 
├── ebin
│   ├── proper.app
│   ├── proper.beam
│   ├── proper_arith.beam
│   ├── proper_array.beam
│   ├── proper_dict.beam
│   ├── proper_erlang_abstract_code.beam
│   ├── proper_fsm.beam
│   ├── proper_gb_sets.beam
│   ├── proper_gb_trees.beam
│   ├── proper_gen.beam
│   ├── proper_gen_next.beam
│   ├── proper_orddict.beam
│   ├── proper_ordsets.beam
│   ├── proper_prop_remover.beam
│   ├── proper_queue.beam
│   ├── proper_sa.beam
│   ├── proper_sets.beam
│   ├── proper_shrink.beam
│   ├── proper_statem.beam
│   ├── proper_symb.beam
│   ├── proper_target.beam
│   ├── proper_transformer.beam
│   ├── proper_types.beam
│   ├── proper_typeserver.beam
│   ├── proper_unicode.beam
│   ├── proper_unused_imports_remover.beam
│   └── vararg.beam
└── include
    ├── proper.hrl
    ├── proper_common.hrl
    ├── proper_internal.hrl
    └── proper_param_adts.hrl
```

This installation doesn't work, the application is not recognised by the Erlang runtime.

```
$ export ERL_LIBS=/opt/homebrew/Cellar/proper/1.4
$ erl
Erlang/OTP 24 [erts-12.1.5] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [dtrace]

Eshell V12.1.5  (abort with ^G)
1> c(test).
test.erl:3:14: can't find include lib "proper/include/proper.hrl"
%    3| -include_lib("proper/include/proper.hrl").
...
```

According to the Erlang documentation (https://www.erlang.org/doc/design_principles/applications.html) the application structure needs to have this naming format: `${application}-${version}`. As a further test I just rsync'd the content in a folder with the correct naming convention and it works perfectly:

```
$ export ERL_LIBS=/Users/miki/proper-1.4
$ erl
Erlang/OTP 24 [erts-12.1.5] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [dtrace]

Eshell V12.1.5  (abort with ^G)
1> c(test).
{ok,test}
```

**Comment by carlocab** [comment]
What happens if the version is omitted? e.g `ERL_LIBS=/Users/miki/proper`?

**Comment by mezzomondo** [comment]
```
$ export ERL_LIBS=/Users/mikifossati/proper
$ erl
Erlang/OTP 24 [erts-12.1.5] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [dtrace]

Eshell V12.1.5  (abort with ^G)
1> c(test).
test.erl:3:14: can't find include lib "proper/include/proper.hrl"
%    3| -include_lib("proper/include/proper.hrl").
```

**Reply by carlocab** [reply]
Do symlinks work? That is, what if `/Users/miki/proper-1.4` were a symlink to `/opt/homebrew/Cellar/proper/1.4`?

**Reply by mezzomondo** [reply]
Yes they do!

**Reply by carlocab** [reply]
Thanks; I opened https://github.com/Homebrew/homebrew-core/pull/89963.

I ended making a directory containing symlinks to the contents of `/opt/homebrew/Cellar/proper/1.4`. I assume the contents of the directory being symlinks should work fine too. Do you mind checking?

**Reply by mezzomondo** [reply]
If you mean something like this, `proper-1.4` being a directory and not a symlink, with both `ebin` and `install` linked underneath:

```
$ ls -la proper-1.4
total 0
drwxr-xr-x    4 miki staff   128 25 Nov 18:20 .
drwxr-xr-x+ 188 miki  staff  6016 25 Nov 18:21 ..
lrwxr-xr-x    1 miki  staff    36 25 Nov 18:20 ebin -> /opt/homebrew/Cellar/proper/1.4/ebin
lrwxr-xr-x    1 miki  staff    39 25 Nov 18:20 include -> /opt/homebrew/Cellar/proper/1.4/include
```

Yes, it works perfectly.

**Reply by carlocab** [reply]
Yes, exactly. Great, Homebrew/homebrew-core#89963 should fix this then.

**Reply by mezzomondo** [reply]
Amazing, thank you very much!
